eigen_cpp/lib/sparse_solver.cpp

```diff
@@ -0,0 +1,86 @@
+#include "sparse_solver.h"
+
+bool report_error(Eigen::ComputationInfo info) {
+    switch(info) {
+    case NumericalIssue:
+      std::cerr << "Solver Failed: Numerical Issue (prerequisites not met)" << std::endl;
+      return true;
+    case NoConvergence:
+      std::cerr << "Solver Failed: Did not converge" << std::endl;
+      return true;
+    case InvalidInput:
+      std::cerr << "Solver Failed: Invalid input" << std::endl;
+      return true;
+    case Success:
+    default:
+      return false;
+    }
+  }
+
+template <class Solver,
+          typename spmat,
+          typename spmat_elt,
+          typename c_spmat>
+spmat& SparseSolver<Solver, spmat, spmat_elt, c_spmat>
+::c_to_eigen(c_spmat* ptr)
+{
+  return *reinterpret_cast<spmat*>(ptr);
+}
+
+template <class Solver,
+          typename spmat,
+          typename spmat_elt,
+          typename c_spmat>
+c_spmat* SparseSolver<Solver, spmat, spmat_elt, c_spmat>
+::eigen_to_c(spmat& ref)
+{
+  return reinterpret_cast<c_spmat*>(&ref);
+}
+
+template <class Solver,
+          typename spmat,
+          typename spmat_elt,
+          typename c_spmat>
+SparseSolver<Solver, spmat, spmat_elt, c_spmat>
+::SparseSolver(c_spmat *c_coefficients, c_spmat *c_observations)
+{
+  spmat row_major_coeffs = c_to_eigen(c_coefficients).pruned();
+  spmat row_major_observ = c_to_eigen(c_observations).pruned();
+
+  coefficients = new col_major_spmat(row_major_coeffs);
+  observations = new col_major_spmat(row_major_observ);
+
+  coefficients->makeCompressed();
+  observations->makeCompressed();
+
+  solver.compute(*coefficients);
+}
+
+template <class Solver,
+          typename spmat,
+          typename spmat_elt,
+          typename c_spmat>
+SparseSolver<Solver, spmat, spmat_elt, c_spmat>
+::~SparseSolver()
+{
+  delete coefficients;
+  delete observations;
+}
+
+template <class Solver,
+          typename spmat,
+          typename spmat_elt,
+          typename c_spmat>
+c_spmat* SparseSolver<Solver, spmat, spmat_elt, c_spmat>::solve()
+{
+  col_major_spmat col_major_solution;
+  spmat*          row_major_solution;
+  c_spmat*        c_solution;
+
+  if(report_error(SparseSolver::solver.info())) { return NULL; }
```

Comment:
It would be nice to have an error type here, and report back the right error. But this can be postponend

Reply:
Agreed, seemed simpler to just go with an option at this time.

Right now, it also prints an error message to the screen, just indicating which of the three error types was reported in solver.info(). Not sure whether you guys want to keep that behavior or not.

Reply:
For now it is fine, at least there is some feedback until we move to proper error reporting